x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/MlJobSnapshotUpgradeIT.java

```diff
@@ -0,0 +1,276 @@
+     */
+    public void testSnapshotUpgrader() throws Exception {
+        hlrc = new HLRC(client()).machineLearning();
+        //assumeTrue("Snapshot upgrader should only upgrade from the last major", UPGRADE_FROM_VERSION.major < 7);
+        Request adjustLoggingLevels = new Request("PUT", "/_cluster/settings");
+        adjustLoggingLevels.setJsonEntity(
+            "{\"transient\": {" +
+                "\"logger.org.elasticsearch.xpack.ml\": \"trace\"" +
+                "}}");
+        client().performRequest(adjustLoggingLevels);
+        switch (CLUSTER_TYPE) {
+            case OLD:
+                createJobAndSnapshots();
+                break;
+            case MIXED:
+                // Add mixed cluster test after backported
+                break;
+            case UPGRADED:
+                ensureHealth((request -> {
+                    request.addParameter("timeout", "70s");
+                    request.addParameter("wait_for_nodes", "3");
+                    request.addParameter("wait_for_status", "yellow");
+                }));
+                testSnapshotUpgrade();
+                waitForPendingUpgraderTasks();
+                break;
+            default:
+                throw new UnsupportedOperationException("Unknown cluster type [" + CLUSTER_TYPE + "]");
+        }
+    }
+
+    private void testSnapshotUpgrade() throws Exception {
+        Job job = getJob(JOB_ID).jobs().get(0);
+        String currentSnapshot = job.getModelSnapshotId();
+
+        GetModelSnapshotsResponse modelSnapshots = getModelSnapshots(job.getId());
+        assertThat(modelSnapshots.snapshots(), hasSize(2));
+        assertThat(modelSnapshots.snapshots().get(0).getMinVersion().major, equalTo((byte)7));
+        assertThat(modelSnapshots.snapshots().get(1).getMinVersion().major, equalTo((byte)7));
+
+        ModelSnapshot snapshot = modelSnapshots.snapshots()
+            .stream()
+            .filter(s -> s.getSnapshotId().equals(currentSnapshot) == false)
+            .findFirst()
+            .orElseThrow(() -> new ElasticsearchException("Not found snapshot other than " + currentSnapshot));
+
+        assertThat(hlrc.upgradeJobSnapshot(
+            new UpgradeJobModelSnapshotRequest(JOB_ID, snapshot.getSnapshotId(), null, true),
+            RequestOptions.DEFAULT).isCompleted(), is(true));
+
+        List<ModelSnapshot> snapshots = getModelSnapshots(job.getId(), snapshot.getSnapshotId()).snapshots();
+        assertThat(snapshots, hasSize(1));
+        assertThat(snapshot.getLatestRecordTimeStamp(), equalTo(snapshots.get(0).getLatestRecordTimeStamp()));
+
+        // Does the snapshot still work?
+        assertThat(hlrc.getJobStats(new GetJobStatsRequest(JOB_ID), RequestOptions.DEFAULT)
+                .jobStats()
+                .get(0)
+                .getDataCounts().getLatestRecordTimeStamp(),
+            greaterThan(snapshot.getLatestRecordTimeStamp()));
```

Comment:
After backport I want to add a mixed cluster test (to make sure the mixed node error throws) and I want to verify that the `min_version` is updated on the new snapshot.

Right now, since 8.x does not support upgrades from 6.x, that is not possible here. But in 7.x, it will be good to test that `min_version` gets adjusted.